### ISSUE TITLE
Report from meeting at AOFD

### ISSUE THREAD

**Issue by rabernat** [pyqg/pyqg#197]
Today we had a lunch meeting at AOFD including myself, @pittwolfe, @worthamc, and @navidcy. Unfortunately @shfer and @MFJansen could not attend. This issue tries to summarize an hour-long discussion.

### things we love about pyqg
- reasonably well documented
- well tested
- simple, useful for teaching
- [mostly] all python
- fast enough
- hackable (easy to plug in new diagnostics, forcing, etc.)
- pleasing social experience working together
- good way for developers to learn about python

### current challenges facing pyqg
- Scientific oversight of the model has lapsed a bit. This is most obviously highlighted by the bug related to the non-uniform barotropic flow uncovered by @worthamc. (Do we have an issue for this yet? If not, please open one.) Such problems should not have made it through review.
- The original developers (myself and @MFJansen) are not especially focused on pyqg right now
- pyqg is not feature-complete for many problems (e.g. #137, #137, #134, #133, #132 etc.)
- There are alternatives out there (Julia, daedalus, etc.)

The fundamental question we tried to answer is **do we freeze pyqg development (no new features) or do we want to continue improving it?**

The general consensus was **we want to keep developing pyqg**. The pros outweigh the cons. However, no one has the time or expertise to do it alone. Therefore, we want to expand the circle of developers and try to work in a more coordinated way.

### action items going forward
- We need to resolve the current bug with the zonally varying background flow. (This may require reverting #159 and starting fresh.) We should first agree on the basic governing equations for non-uniform background states and the algorithm we wish to use. (The same issues have to be resolved for bottom topography; #133).
- Once this is done, we can collectively agree on how the kernel should look: which terms are functions of space vs constants, etc.
- We should work to formally define the core kernel API specification independently of its implementation. 
- With this spec in hand, we should refactor the kernel, in the process making it easier to implement new kernels with different approaches / numerics (e.g. gpu kernel #167). 
- We should improve the unit testing of the kernel and worry less about exact numerical agreement in the integration tests.

We all agreed we would love to have input from @shfer regarding the optimal numerical methods. This will help us refine the kernel specification.

This is a lot of work, but if we work together it should be manageable. (If anyone else is watching this, we welcome your involvement.)

### remaining questions
- Do we want to allow different methods to the vertical dimension (e.g. spectral in the vertical?) If so, what changes to the kernel does this require.
- Who can commit to handling the "maintenance" of pyqg (responding to issues, fixing bugs, making sure CI and docs are working, cutting new releases, etc.) on a day-to-day basis? It should probably be someone other than myself and @MFJansen, since we have become a bottleneck.
- What are our highest priority features, after topography and non-uniform background flow.

Please feel free to add anything else I forgot.

**Comment by rabernat** [comment]
From @MFJansen via email:

- I think our highest priority right now should be to fix issues in the existing code and documentation. There are a lot of open issues right now.
- Of course anybody who wants to implement a new feature is encouraged to do so, but I think we should make sure that additions are properly tested and don’t break anything existing (highest priority), but also are correct in the feature they add (e.g. we may want to require an example/test where we know the correct answer)
- I personally won’t be able to spend much time working on pyQG in the next few months, and do not have anyone in my group who will be working on it. To the degree that I will be able to contribute, I will focus on dealing with existing issues and errors.  Specifically, I will talk to @mbueti about the zonally varying background flow implementation, which I believe is buggy (beyond the issues that have been pointed out already).

**Comment by MFJansen** [comment]
Just to clarify that I wrote the above before reading this meeting summary, so it's not a "response". Anyway, it sounds like we mostly agree.

**Comment by rabernat** [comment]
I agree with you :100:% about fixing existing issues before undertaking new development.

**Comment by shfer** [comment]
I appreciate being asked to contribute, but given my plodding python abilities, please don't expect much!  Regarding what you've summarized above, however, I do have a couple *opinions* that might be useful:

* Adding topography is far more important than meridional variation in the background flow.  The latter, I think, pushes the boundaries of what one can reasonably get out of doubly-periodic QG simulations.  There are definitely interesting questions there, but I think it would be used far less often than topography.  

* Regarding the vertical grid options, I have many thoughts about this.  First, if you want to resolve the surface stuff in QG, you need a very fine grid near the surface.  But non-uniform grids degrade the standard centered difference (or layer-differencing) method from O(dz^2) to O(dz).  There are compact methods that allow non-uniform gridding that could be added.  I had a conversation with Jeffrey Early about this last year, and he sent the attached paper [Fornberg1998.pdf](https://github.com/pyqg/pyqg/files/3364075/Fornberg1998.pdf)
Regarding spectral options, based on work with Ian Grooms, I suspect Chebychev methods will be superior, though at the loss of exact energy conservation.  But here I think it may be hard to compete with Dedalus. I plan to try to get that going in a QG configuration that resolves boundary dynamics this summer.  Scott Bachman is trying to do something similar.  
* I agree that allowing for GPU kernel should be a high-priority.  Having seen a demonstration with python and fortran code using FFTs in a fluid model, the speedup was kind of mind-blowing.  

If there is something specific any of you think I could be of help with, please do ask.  BTW, have you invited Jeffrey Early to be involved with this effort?  He has done a lot of work already on QG development, and he's a bonafide software developer  ..  he could be very helpful, if convinced.

**Comment by francispoulin** [comment]
I hope everyone is enjoying the meeting and sorry that I am not there to join in the fun.

I have been an Associate Dean, Undergraduate Studies for the last two years and that has taken up pretty much all of my time.  But I am now done with that and getting back to research and am keen to contribute more to pyqg.  I know that pyqg has grown since last I used it so I will do a fresh install and play with it over the next week and learn more where it's at.  

I am interested in contributing.  

Below are a few not-so-random thoughts.

- I agree that topography is is probably more important than mean flows, and should be easier to implement and a good place to start.  I would be nice to get mean flows working afterwards though.
- I have been using Chebyschev grids since my PhD and can't say enough good things above them.  It would be a great way of better resolving both the surface and bottom dynamics.  True that they won't necessarily conserve energy exactly, but neither do Fourier spectral methods and they still work beautifully.
- I clearly need to brush up on Dedalus as well.
- If one wanted to accurately resolve surface dynamics then having Surface QG would be good to include as well. I believe we had something a while ago but not sure where things stand right now. I will learn soon.

On a very different note, because someone mentioned Julian, I wanted to mention that I'm getting students this semester to translate the code from "Spectral Methods in Matlab" by Nick Trefethen (where I learned the joys of cheb) into Julia.  I'm hoping to post this somewhere after we are done but will ask Nick what he thinks before I do anything.  Not directly related to this but I mention it in case anyone is interested.

Cheers, Francis

**Comment by navidcy** [comment]
@francispoulin, I presume you aware of [ApproxFun.jl](https://github.com/JuliaApproximation/ApproxFun.jl), which is sort of the Julia-implementation of Trefethen's et al Chebfun?

Regarding @shfer's comment on vertical discretization: If one envisions to take `pyqg` beyond the few-layered vertical discretization, then indeed Chebychev methods seem to be the best wav forward. But at that point the complexity goes beyond the intention of simple QG doubly-periodic models and, I feel, that one should simply use [Dedalus](http://dedalus-project.org).
